### ISSUE TITLE
valid-jsdoc: optional properties

### ISSUE THREAD

**Issue by monolithed** [eslint/eslint#6745]
```js
{
	/**
	 * @property {string} [provider]
	 * @property {Array} [features]
	 */
	get ({ provider, features = [] }) { }
}
```

> ✘  http://eslint.org/docs/rules/valid-jsdoc  JSDoc syntax error  

```js
{
	/**
	 * @property {string} provider
	 * @property {Array} features
	 */
	get ({ provider, features = [] }) { }
}
```

> ✓

How to mark optional properties?

**Comment by lukeapage** [comment]
just guessing but maybe it supports the google closure syntax?

```
{
    /**
     * @property {string=} provider
     * @property {Array=} features
     */
    get ({ provider, features = [] }) { }
```

http://usejsdoc.org/tags-param.html

**Comment by monolithed** [comment]
https://closure-compiler.appspot.com

```js
// ==ClosureCompiler==
// @compilation_level SIMPLE_OPTIMIZATIONS
// @warning_level VERBOSE
// @language ECMASCRIPT6
// @output_file_name default.js
// ==/ClosureCompiler==


let object = {
    /**
     * @property {string} [provider]
     * @property {Array} [features]
     */
    get ({ provider, features = [] }) { }
}

object.get({});
```

> JSC_POSSIBLE_INEXISTENT_PROPERTY: Property provider never defined on $jscomp$destructuring$var1 at line 6 character 11
    get ({ provider, features = [] }) { }
           ^
JSC_POSSIBLE_INEXISTENT_PROPERTY: Property features never defined on $jscomp$destructuring$var1 at line 6 character 21
    get ({ provider, features = [] }) { }


```js
// ==ClosureCompiler==
// @compilation_level SIMPLE_OPTIMIZATIONS
// @warning_level VERBOSE
// @language ECMASCRIPT6
// @output_file_name default.js
// ==/ClosureCompiler==


let object = {
    /**
     * @param {string=} provider
     * @param {Array=} features
     */
    get ({ provider, features = [] }) { }
}

object.get({});
```

> JSC_INEXISTENT_PARAM: parameter features does not appear in object.get's parameter list at line 6 character 4
    get ({ provider, features = [] }) { }
    ^
JSC_INEXISTENT_PROPERTY: Property provider never defined on String at line 6 character 11
    get ({ provider, features = [] }) { }
           ^
JSC_INEXISTENT_PROPERTY: Property features never defined on String at line 6 character 21
    get ({ provider, features = [] }) { }
                     ^
JSC_TYPE_MISMATCH: actual parameter 1 of object.get does not match formal parameter
found   : {}
required: (string|undefined) at line 10 character 11
object.get({});

**Comment by eslintbot** [comment]
Hi @monolithed, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by platinumazure** [comment]
@monolithed You linked to the closure compiler standard- just want to note that we use Doctrine to parse JSDoc comments, and Doctrine does not claim to support all of the Closure Compiler documentation constructs. You may wish to read up on Doctrine to see if your use case is supported.

**Comment by monolithed** [comment]
I have not found the right way to describe the optional properties, so I close this issue. Feel free to reopen.

**Comment by hashchange** [comment]
@monolithed An optional `@property` can't be annotated like a `@param`, so you can't use 
```
@property {string} [provider]
```
or 
```
@property {string=} provider
```
Apparently, there isn't any special notation to mark a property as optional.

But an optional property is really nothing more than a property which can also be left undefined. So all you have to do is to declare it as 
```
@property {string|undefined} provider
```
